### ISSUE TITLE
Issue with Base64 encoding of ID Token

### ISSUE THREAD

**Issue by MrLinio** [juanifioren/django-oidc-provider#115]
Hi,
A client of us reported an issue with the encoding of the id_token : 

> {"iss":"https://sfival.bcsd-nnrgus.com/openid","iat":1471857896,"exp":1471858496,"aud":"pamidea_val","sub":"5444","auth_time":1471857586}

is transformed : 

> eyJpc3MiOiJodHRwczovL3NmaXZhbC5iY3NkLW5ucmd1cy5jb20vb3BlbmlkIiwiaWF0IjoxNDcxODU3ODk2LCJleHAiOjE0NzE4NTg0OTYsImF1ZCI6InBhbWlkZWFfdmFsIiwic3ViIjoiNTQ0NCIsImF1dGhfdGltZSI6MTQ3MTg1NzU4Nn0

But because of the number of bytes of the string, the Base64 stipulates a = must be put at the end in order for the encoding to be valid.

**Comment by juanifioren** [comment]
Hi thanks for reporting. I'll try to figure out what is the problem with the encoding. We use a function of `pyjwkest` lib. So we have dependency there.

[The function in charge of the encoding](https://github.com/juanifioren/django-oidc-provider/blob/master/oidc_provider/lib/utils/token.py#L65)

Any help will be great.

**Comment by MrLinio** [comment]
Hi.

Hopefully the answer system works (I'm not a big user of github ;)).
I did a little digging this morning, and I think i found the issue :

https://github.com/rohe/pyjwkest/blob/master/src/jwkest/__init__.py

Line 145 is this :
return base64.urlsafe_b64encode(b).rstrip(b"=")

Which is eventually the line that is used to encode the JWT.
Although i love going through RFC's and such, although i'm really unsure
that I raised an actual issue...

Ok, so I went through the RFC :

A JWT is represented as a sequence of URL-safe parts separated by
   period ('.') characters.  Each part contains a base64url-encoded
   value.  The number of parts in the JWT is dependent upon the
   representation of the resulting JWS using the JWS Compact
   Serialization or JWE using the JWE Compact Serialization.


I do undertand that the client made a fool of myself, hopefully you'll read
it the same way?

All the libraries I saw there seems to point to the same conclusion.
https://jwt.io/

Well.
I guess I made you lose your time, but it was one of those time, I needed
someone to listen to me to help me discover the issue, so sorry about that.
So the only thing left for me to say is : Thanks, your library was
tremendously helpful to what we needed to achieve on our end!



2016-08-23 18:34 GMT+02:00 Juan Ignacio Fiorentino <notifications@github.com
